fix: memory leak on connection crashes

hpopp opened the pull request:
On a per-connection basis, everything is now supervised under a ConnectionSupervisor. Hpack tables and streams were previously getting orphaned if Connection crashed.

hpopp committed:
feat: better supervision structure

hpopp committed:
Merge branch 'master' into new-sup

hpopp committed:
chore: bumped kadabra to v0.3.5

Also: better test coverage

talklittle commented on lib/connection_supervisor.ex:
```diff
@@ -0,0 +1,43 @@
+  use Supervisor
+  import Supervisor.Spec
+
+  alias Kadabra.{Connection, Hpack, Stream}
```
warning: unused alias Connection

talklittle commented on lib/kadabra.ex:
```diff
@@ -2,7 +2,7 @@ defmodule Kadabra do
   HTTP/2 client for Elixir.
   """
-  alias Kadabra.{Connection}
+  alias Kadabra.{Connection, ConnectionSupervisor}
```
warning: unused alias Connection

talklittle commented:
~~I'm trying out this branch, will monitor over the coming days and let you know how it goes.~~

**EDIT:** Nevermind, forgot that it needs Pigeon changes (`max-stream-fix` branch presumably) in addition. That branch looks like it's not 100% finished, so I'll hold off testing for now.

hpopp committed:
feat: entirely new supervision strategy

Kadabra.Supervisor starts a ConnectionSupervisor and StreamSupervisor.
ConnSupervisor manages encoder/decoder tables and the Connection worker
with a :one_for_all restart strategy. StreamSupervisor manages the
streams with :one_for_one restart.

To keep the API the same, Kadabra.Supervisor.start_link returns the pid
of the supervisor itself. The Connection worker is keyed in the registry
by this supervisor pid. The user still makes calls with the returned
pid, but it's forwarded with a via_tuple to the proper worker. Unsure
if this might impact performance.

Possible TODO: Consider keying everything by supervisor pid instead of
the generated ref. While conflicts should be virtually nonexistent, this
would guarantee it.

hpopp commented:
After spending all day working through this, I think I've come up with a proper supervision strategy. Commit message explains the structure, but the basic gist: Kadabra.Connection is keyed in the registry by the overall parent supervisor, whose pid is returned from `Kadabra.open/2`. When making requests, the client uses the pid as normal, but it's actually forwarded with a registry `:via` tuple.

The solution feels hacky, but it appears to work just fine. Thoughts?

talklittle commented:
* Using supervisor pids makes perfect sense to me, assuming they are cleaned up properly, i.e. the client gets notified when supervisors die, so clients won't try to use dead pids. But I guess nothing has changed really, since referencing a dead stream ref would have had the same result, right?
* I guess the "hacky" part is alluded to by your "possible TODO" in the commit message? That it's now passing around a pid in addition to a now-possibly-redundant "ref"? (Sorry I haven't dived into the code enough to know how else "ref" is used)  
If so, then I think it's a good refactor to do at some point, but certainly not a blocker for this branch.
* Re: Performance impact, is the additional registry lookup only when the end-user calls into Kadabra? Do you have an example scenario in mind where an end-user is calling into Kadabra frequently enough that performance matters here?  
Actually in the [Registry docs](https://hexdocs.pm/elixir/Registry.html#module-registrations) it mentions that "Looking up, dispatching and registering are efficient and immediate at the cost of delayed unsubscription." so the lookup penalty can probably be ignored.  
~~As for the "delayed unsubscription," I *think* that caveat maybe only applies to a partitioned/distributed Registry, so hopefully irrelevant for Kadabra.~~ **EDIT:** Nevermind, Registry is local-only, not distributed. I guess it's referring to async Registry operations that might do a lookup at the beginning of the operation, then the process dies, yet the async operation still returns the process as though it were alive.

hpopp commented:
The ref is used to identify all of the workers in the registry except `Connection`. Ideally `Kadabra.open/2` would return this ref instead of a pid, but I'm not sure I've seen this convention used anywhere (not to mention it'd break API). It feels hacky to me because in this case, you're using the supervisor pid like a ref, even though its functionally no different. That said, I'm not sure the end-user would even notice or care.

Assuming this approach makes sense, every worker should instead be identified by the pid instead of the ref, which will be relatively easy to change.

It's good that performance should remain effectively the same. I'd forgotten registry lookups were fast.

hpopp committed:
chore: update CHANGELOG

hpopp committed:
fix: removed unused ConnectionSupervisor.start_connection
